### DISCUSSION TITLE
What is the best way to test the pyxel supported (I believe v0.23) version out?

### DISCUSSION THREAD

**Discussion by ChrisKahrs** [pyscript/pyscript#1367] [Packages]
Is there a build of this in unstable yet?  Or should I wait until there is?   Once it is there (or a branch) how is best to call it from pyscript.com or the like...
    <script defer src="https://pyscript.net/unstable/pyscript.js"></script> or a direct branch call? 

Love the work so far, really impressive.

Thanks a ton!

**Comment by JeffersGlass** [comment]
Pyodide 0.23.0 isn't _officially_ supported in PyScript yet (there's [an open PR by yours truly](https://github.com/pyscript/pyscript/pull/1347) trying to get it supported). Once that's merged, the Pyxel integration will be available in unstable. 

That said, Pyodide 0.23.0 _mostly_ works with the current 2023.03.1 release. If you just want to try out the Pyxel support, you can do so by pointing the `[[interpreters]]` section of \<py-config\> to the new Pyodide release (in your own site or PyScript.com). See below for a full example.

Fair warning - as you've probably seen, support for this is still experimental on the Pyodide side. For instance, `pyodide_js._api._skip_unwind_fatal_error = True` is basically telling Pyodide to ignore a normally fatal error caused by how Emscripten handles enterring infinite loops (or "simulated" infinite loops). See [the original PR](https://github.com/pyodide/pyodide/pull/3508#issuecomment-1488837896) for more details. So all in all, this works, but the edges are still being sanded off.

```html
<!-- index.html -->
<head>
    <script defer src="https://pyscript.net/releases/2023.03.1/pyscript.js"></script>
    <link rel="stylesheet" href="https://pyscript.net/releases/2023.03.1/pyscript.css">
</head>
<body>
    <py-config>
        packages=['pyxel']
        [[interpreters]]
        src = "https://cdn.jsdelivr.net/pyodide/v0.23.0/full/pyodide.js"
        name = "pyodide-0.23.0"
        lang = "python"
    </py-config>
    <py-script src="pyxel_demo.py"></py-script>
    <canvas id="canvas" tabindex="-1"></canvas>
</body>
```
```python
#pyxel_demo.py

import js
import pyodide
import pyodide_js

import pyxel

# Tell Pyodide/Emscripten where to draw the output
pyodide_js.canvas.setCanvas2D(js.document.getElementById("canvas"))

# Suppress a fatal error - running the Pyxel infinite loop leaves dead frames on the call stack;
# This supresses this warning for now. See https://github.com/pyodide/pyodide/pull/3508#issuecomment-1488837896
pyodide_js._api._skip_unwind_fatal_error = True

# Pyxel requires some additional setup for its onscreen gamepad before running
# Haven't personally figured out why it's necessary to set manually
pyodide.code.run_js(
    """_virtualGamepadStates = [
        false, // Up
        false, // Down
        false, // Left
        false, // Right
        false, // A
        false, // B
        false, // X
        false, // Y
    ];
    """
)

# Now you're set up to run your aplication code.
# A very simple Pyxel game. Press up and down to adjust the color of the text

class App():
    def __init__(self): 
        self.color = 1
        pyxel.init(256, 256)
        pyxel.run(self.update, self.draw)

    def draw(self):
        pyxel.cls(0)
        for y in range(0, 256, 16):
            pyxel.text(0, y, "Hello world!", self.color)
    
    def update(self):
        if pyxel.btnp(pyxel.KEY_UP):
            self.color = (self.color + 1) % 15 + 1
        if pyxel.btnp(pyxel.KEY_DOWN):
            self.color = (self.color + 13) % 15 + 1
        
App()
```

**Reply by hoodmane** [reply]
I am personally very worried about the behavior enabled by
```py
pyodide_js._api._skip_unwind_fatal_error = True
```
In fact, I strongly recommend that nobody should actually do this (except for mischief-related purposes).

The fatal error that it disables doesn't corrupt any memory, but it allows a mismatch between the Python stack trace and the C stack trace. There are likely interpreter invariants that expect that the Python stack and the C stack are in agreement. It may never cause any problems, but if it does cause problems they can be arbitrarily weird.

From a more practical perspective, since this is supposed to run as a render loop, the fact that it leaks all local variables and doesn't run any finalizers means that it is doomed to run out of memory even if it doesn't fail in weirder ways.

**Comment by ChrisKahrs** [comment]
Thanks @JeffersGlass, much appreciated the work on these pieces!
